### PR TITLE
Feature/sg-43: get spotify recommendations with ML results, return mapped data

### PR DESCRIPTION
**Pull request by patkarns**
Endpoint: `http://localhost:4000/spotify/recs/access_token`

Mapped data should return an array of tracks (`Tracks[]`). Will update Swagger once this is finalized
```
export interface Track {
        id: string;
        imageUrl: string | null;
        name: string;
        artists: Artist[] | null;
        duration: string;
} 

export interface Artist {
    name: string;
    id: string;
    spotify_url: string;
}
```

Example response:
```
[
    {
        "id": "519uJbE3zyKLlToVA65PrP",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273f5eaf1b4d675b763303a127b",
        "name": "Counting Stars",
        "artists": [
            {
                "name": "OneRepublic",
                "id": "5Pwc4xIPtQLFEnJriah9YJ",
                "spotify_url": "https://open.spotify.com/artist/5Pwc4xIPtQLFEnJriah9YJ"
            }
        ],
        "duration": "4m 17s"
    },
    {
        "id": "09PziVJjD74IagSfdN58wH",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b27313495130dccead15dfd71cfb",
        "name": "Confidant",
        "artists": [
            {
                "name": "BabyJake",
                "id": "07Asx51VtMw5kbNXKrpZlq",
                "spotify_url": "https://open.spotify.com/artist/07Asx51VtMw5kbNXKrpZlq"
            }
        ],
        "duration": "3m 2s"
    },
    {
        "id": "6cJqlSO19vsfCsH1FeFI35",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b2730193b418a25d229453c1927b",
        "name": "Santeria",
        "artists": [
            {
                "name": "Sublime",
                "id": "0EdvGhlC1FkGItLOWQzG4J",
                "spotify_url": "https://open.spotify.com/artist/0EdvGhlC1FkGItLOWQzG4J"
            }
        ],
        "duration": "3m 3s"
    },
    {
        "id": "6j5J40LiUKW1roHvlBcN5s",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273a172f60f5608561ec9b9abbd",
        "name": "Chaparral",
        "artists": [
            {
                "name": "Healy",
                "id": "2Yhge9MsE7qKcV0eWsuuHM",
                "spotify_url": "https://open.spotify.com/artist/2Yhge9MsE7qKcV0eWsuuHM"
            }
        ],
        "duration": "3m 48s"
    },
    {
        "id": "7gNRYLOwgmzMEpgXARSnX9",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273c12636cecde49870f08d6ecb",
        "name": "Leave Me (feat. Marc E. Bassy)",
        "artists": [
            {
                "name": "Skizzy Mars",
                "id": "00Z3UDoAQwzvGu13HoAM7J",
                "spotify_url": "https://open.spotify.com/artist/00Z3UDoAQwzvGu13HoAM7J"
            },
            {
                "name": "Marc E. Bassy",
                "id": "3tQx1LPXbsYjE9VwN1Peaa",
                "spotify_url": "https://open.spotify.com/artist/3tQx1LPXbsYjE9VwN1Peaa"
            }
        ],
        "duration": "2m 51s"
    },
    {
        "id": "0h1YzxN6QTjEkJGU6fxsyt",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273d92afcf66072c83e4b28c6bb",
        "name": "I'm Ready To Move On/Mickey Mantle Reprise",
        "artists": [
            {
                "name": "Bleachers",
                "id": "2eam0iDomRHGBypaDQLwWI",
                "spotify_url": "https://open.spotify.com/artist/2eam0iDomRHGBypaDQLwWI"
            }
        ],
        "duration": "4m 23s"
    },
    {
        "id": "0KohaUnLZU0PNJC3RZYDZ0",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b27367d5041ebaa146ce6123e043",
        "name": "La Vie En Rose",
        "artists": [
            {
                "name": "Lucy Dacus",
                "id": "07D1Bjaof0NFlU32KXiqUP",
                "spotify_url": "https://open.spotify.com/artist/07D1Bjaof0NFlU32KXiqUP"
            }
        ],
        "duration": "2m 50s"
    },
    {
        "id": "1wAXODAAL6hY64ZdhrnjBO",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273eae3a0656610821e43139c10",
        "name": "Drive By",
        "artists": [
            {
                "name": "Train",
                "id": "3FUY2gzHeIiaesXtOAdB7A",
                "spotify_url": "https://open.spotify.com/artist/3FUY2gzHeIiaesXtOAdB7A"
            }
        ],
        "duration": "3m 16s"
    },
    {
        "id": "6chdRBWviHlm7JAtwgflBP",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b2732231bc8cc3e2cd143e95ea38",
        "name": "Maps",
        "artists": [
            {
                "name": "Maroon 5",
                "id": "04gDigrS5kc9YWfZHwBETP",
                "spotify_url": "https://open.spotify.com/artist/04gDigrS5kc9YWfZHwBETP"
            }
        ],
        "duration": "3m 9s"
    },
    {
        "id": "2dBwB667LHQkLhdYlwLUZK",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273cb76604d9c5963544cf5be64",
        "name": "Summertime Sadness",
        "artists": [
            {
                "name": "Lana Del Rey",
                "id": "00FQb4jTyendYWaN8pK0wa",
                "spotify_url": "https://open.spotify.com/artist/00FQb4jTyendYWaN8pK0wa"
            }
        ],
        "duration": "4m 25s"
    },
    {
        "id": "6LN15vwkSE1ZiA8IdBjl7k",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273e3ddc84da0e77626a8ff6e26",
        "name": "ONE PUNCH",
        "artists": [
            {
                "name": "Aries",
                "id": "3hOdow4ZPmrby7Q1wfPLEy",
                "spotify_url": "https://open.spotify.com/artist/3hOdow4ZPmrby7Q1wfPLEy"
            }
        ],
        "duration": "2m 25s"
    },
    {
        "id": "6YArGkJUr82ehzRFa2YaRK",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b2734849278ce9876ebea7353d66",
        "name": "Wake Me",
        "artists": [
            {
                "name": "Bleachers",
                "id": "2eam0iDomRHGBypaDQLwWI",
                "spotify_url": "https://open.spotify.com/artist/2eam0iDomRHGBypaDQLwWI"
            }
        ],
        "duration": "2m 43s"
    },
    {
        "id": "1iXBApi39l5r6lJj9WEXXS",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b2731ff513d240271aa9976f021f",
        "name": "Hideaway",
        "artists": [
            {
                "name": "Kiesza",
                "id": "4zxvC7CRGvggq9EWXOpwAo",
                "spotify_url": "https://open.spotify.com/artist/4zxvC7CRGvggq9EWXOpwAo"
            }
        ],
        "duration": "4m 11s"
    },
    {
        "id": "0KpfYajJVVGgQ32Dby7e9i",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273706a70e4bfb15d33b8ec74a1",
        "name": "Hey, Soul Sister",
        "artists": [
            {
                "name": "Train",
                "id": "3FUY2gzHeIiaesXtOAdB7A",
                "spotify_url": "https://open.spotify.com/artist/3FUY2gzHeIiaesXtOAdB7A"
            }
        ],
        "duration": "3m 36s"
    },
    {
        "id": "2bzUVEvpZ7At5cYz1kOLI9",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273667f8cfd1be0d0cc2b825e25",
        "name": "Gasoline (feat. Taylor Swift)",
        "artists": [
            {
                "name": "HAIM",
                "id": "4Ui2kfOqGujY81UcPrb5KE",
                "spotify_url": "https://open.spotify.com/artist/4Ui2kfOqGujY81UcPrb5KE"
            },
            {
                "name": "Taylor Swift",
                "id": "06HL4z0CvFAxyc27GXpf02",
                "spotify_url": "https://open.spotify.com/artist/06HL4z0CvFAxyc27GXpf02"
            }
        ],
        "duration": "3m 13s"
    },
    {
        "id": "44psOy0D0SP8rcIiUgKgBs",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b2737f768e99c363b6ee0995556e",
        "name": "Tightrope",
        "artists": [
            {
                "name": "WALK THE MOON",
                "id": "6DIS6PRrLS3wbnZsf7vYic",
                "spotify_url": "https://open.spotify.com/artist/6DIS6PRrLS3wbnZsf7vYic"
            }
        ],
        "duration": "3m 29s"
    },
    {
        "id": "5Sf3GyLEAzJXxZ5mbCPXTu",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273067dfba04c23ce2887041fff",
        "name": "Waves - Robin Schulz Radio Edit",
        "artists": [
            {
                "name": "Mr. Probz",
                "id": "33W1pnW9zScZtYTnAoWnOT",
                "spotify_url": "https://open.spotify.com/artist/33W1pnW9zScZtYTnAoWnOT"
            },
            {
                "name": "Robin Schulz",
                "id": "3t5xRXzsuZmMDkQzgOX35S",
                "spotify_url": "https://open.spotify.com/artist/3t5xRXzsuZmMDkQzgOX35S"
            }
        ],
        "duration": "3m 28s"
    },
    {
        "id": "6Hx7kdHB9kGdxnw2EwSx2V",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b27399650106545ea6d32b63fea2",
        "name": "Million Bucks",
        "artists": [
            {
                "name": "Smallpools",
                "id": "4iiQabGKtS2RtTKpVkrVTw",
                "spotify_url": "https://open.spotify.com/artist/4iiQabGKtS2RtTKpVkrVTw"
            }
        ],
        "duration": "3m 59s"
    },
    {
        "id": "1olNHIIVl4EVwIEPGYIR7G",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b273d7f38f1fc6d03b0f177c93cd",
        "name": "Holy Grail",
        "artists": [
            {
                "name": "JAY-Z",
                "id": "3nFkdlSjzX9mRTtwJOzDYB",
                "spotify_url": "https://open.spotify.com/artist/3nFkdlSjzX9mRTtwJOzDYB"
            },
            {
                "name": "Justin Timberlake",
                "id": "31TPClRtHm23RisEBtV3X7",
                "spotify_url": "https://open.spotify.com/artist/31TPClRtHm23RisEBtV3X7"
            }
        ],
        "duration": "5m 38s"
    },
    {
        "id": "0MuqFuh4MerIiFsYWuEFaP",
        "imageUrl": "https://i.scdn.co/image/ab67616d0000b2732f9cfcc3876eaa0696d5d5c2",
        "name": "Do You There (feat. Marc E. Bassy)",
        "artists": [
            {
                "name": "Skizzy Mars",
                "id": "00Z3UDoAQwzvGu13HoAM7J",
                "spotify_url": "https://open.spotify.com/artist/00Z3UDoAQwzvGu13HoAM7J"
            },
            {
                "name": "Marc E. Bassy",
                "id": "3tQx1LPXbsYjE9VwN1Peaa",
                "spotify_url": "https://open.spotify.com/artist/3tQx1LPXbsYjE9VwN1Peaa"
            }
        ],
        "duration": "4m 47s"
    }
]
```

### CODE REVIEW

File: backend/Node/routes/spotifyRoute.ts [hunk 1]
```diff
@@ -159,4 +170,102 @@ router.get('/hello', async (req: any, res: any) => {
...
+  // TODO: move thiese endpoints somewhere nice, ML server port num should be set and loaded
+  const MLServerRes = await axios.post(`http://localhost:5000/recommend_playlist`, {
+	location,
+	pop,
+	clouds,
+	temp,
+	mood,
+	activity,
+	audio_features,
+	popularity
```
Comment (LukeLNH) [inline backend/Node/routes/spotifyRoute.ts:245]: popularity is already a field inside of audio_features (which is also what the ML server expects), so there's no need to send back popularity on it's own as well :)

**Review by LukeLNH — COMMENTED** [review]
Won't affect performance, but it's probably a good idea to remove the popularity array from the post request to the ML server, since popularity is already apart of audio_features

Other than that I understand nothing so Imma just slap the lgtm here lol 😆

**Comment by hungnguyen3** [comment]
> 

I fixed the popularity array, no more of that now :) Thanks, Long (qua Thanh Long? lol)

File: backend/Node/routes/spotifyRoute.ts (hunk 1 above)
Comment (hungnguyen3) [reply backend/Node/routes/spotifyRoute.ts]: Yeah, I fixed it, no more popularity alone. All in audio_features array

File: backend/Node/routes/route-helpers/spotify-mapper.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            name: a.name,
+            id: a.id,
+            spotify_url: a.external_urls.spotify
+
+        })),
+        duration: `${duration.minutes()}m ${duration.seconds()}s`
+    }
+}
+
+export async function getSeedArtistIdsFromTopTracks(trackIds: string[], access_token: string, topXArtist: number = 2): Promise<string[]> {
```
Comment (hungnguyen3) [inline backend/Node/routes/route-helpers/spotify-mapper.ts:56]: Damn! The interface and detailed types. Too pro! Can't handle this

**Review by hungnguyen3 — APPROVED** [review]
Ran & tested with the ML server! Returns the expected result! Added to Swagger.

**Pull request merged**